### PR TITLE
cmake generate mavlink headers

### PR DESCRIPTION
**Pull request by JacobCrabill**
**Describe problem solved by this pull request**
Especially for new developers, learning how to modify MAVLink message definitions, and then what to do from there (clone a fork of the C library output, commit to that fork, update submodules, etc. etc.) is quite daunting.  Additionally, the C header library is entirely auto-generated, and in general, it's not a great idea to store machine-generated outputs in a version control system (you should instead version control the source of truth, and the tool used to generate the output).

**Describe your solution**
This PR replaces the MAVLink C header library submodule with the "source of truth" MAVLink repo and uses a very simple script to generate the header library from within the CMake build system.

**Other things to note**
This adds a few small requirements to Python and Ubuntu; I believe I've captured these in the `Tools/setup/ubuntu.sh` and `Tools/setup/requirements.txt` files, but there are likely other changes needed elsewhere (for example: the CI system and/or its Docker containers).

### CODE REVIEW

File: src/modules/mavlink/CMakeLists.txt [hunk 1]
```diff
@@ -31,7 +31,18 @@
...
-px4_add_git_submodule(TARGET git_mavlink_v2 PATH "${PX4_SOURCE_DIR}/mavlink/include/mavlink/v2.0")
+px4_add_git_submodule(TARGET git_mavlink_v2 PATH "${PX4_SOURCE_DIR}/mavlink/mavlink")
+
+set(MAVLINK_DIALECT "development")
+set(MAVLINK_GIT_DIR "${PX4_SOURCE_DIR}/mavlink/mavlink")
+set(CLIBRARY_OUT_DIR "${PX4_SOURCE_DIR}/mavlink/include/mavlink/v2.0")
+
+add_custom_command(OUTPUT px4_mavlink_gen.stamp
+    COMMAND "${PX4_SOURCE_DIR}/Tools/mavlink_generate_dialect.sh" ${MAVLINK_DIALECT} "${MAVLINK_GIT_DIR}" "${CLIBRARY_OUT_DIR}" ${PYTHON_EXECUTABLE}
+    COMMAND ${CMAKE_COMMAND} -E touch px4_mavlink_gen.stamp
```
Comment (dagar) [inline src/modules/mavlink/CMakeLists.txt:42]: For convenience I wonder if this should have actual dependencies on common.xml, etc?

File: Tools/mavlink_generate_dialect.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#     assigned to 'common' will be reset to 0, breaking its usage by 'standard'
+#     in the process
+#
+# - In summary: We are ONLY producing the specific dialect requested here
+#   (and its immediate dependencies).
+#
+# Author: Jacob Crabill <jacob@flyvoly.com>
+
+function generate_headers() {
+$PYTHON_EXECUTABLE $MAVLINK_GIT_PATH/pymavlink/tools/mavgen.py \
```
Comment (dagar) [inline Tools/mavlink_generate_dialect.sh:50]: Why not simply call mavgen.py directly from the custom command and have dependencies on the specific xml and output?

Comment (JacobCrabill) [reply Tools/mavlink_generate_dialect.sh:50]: No reason... This is just how I initially got it working (from the script in the ma link repo) and I just went with it.

File: src/modules/mavlink/CMakeLists.txt (hunk 1 above)
Comment (JacobCrabill) [reply src/modules/mavlink/CMakeLists.txt:42]: What do you mean?

Comment (dagar) [reply src/modules/mavlink/CMakeLists.txt]: Updated.

**Comment by dagar** [comment]
Does this work for everyone?

**Review by jlecoeur — APPROVED** [review]
Great improvement.

I have nothing to say as far as PX4 is concerned. However, for other projects that need to communicate with PX4 with the same dialect, the same boilerplate build recipe will be needed. Longer term it may be better to move this to a CMakeList in the Mavlink repo.

File: .gitmodules [hunk 3]
```diff
@@ -66,3 +62,6 @@
 [submodule "Tools/simulation-ignition"]
 	path = Tools/simulation-ignition
 	url = https://github.com/Auterion/px4-simulation-ignition.git
+[submodule "external/mavlink"]
```
Comment (bkueng) [inline .gitmodules:65]: Why not keep it under `mavlink/`?

**Review by bkueng — COMMENTED** [review]
Certainly a welcome simplification.

How is the build time affected? Did you check that incremental build works when changing the xml?

File: .gitmodules (hunk 3 above)
Comment (dagar) [reply .gitmodules:65]: At the time I was fighting incremental submodule move non-sense on top the first commit and it was mildly convenient to use a different location (but no real reason).

I think my preference would be we either bury the submodule as close to the actual usage as possible (eg src/modules/mavlink/mavlink) or keep them together, but off to the side like in `external`.

**Comment by dagar** [comment]
>However, for other projects that need to communicate with PX4 with the same dialect, the same boilerplate build recipe will be needed. Longer term it may be better to move this to a CMakeList in the Mavlink repo.

@jlecoeur if there's enough interest in settling this now I can take a quick look at making `mavlink/mavlink` a proper cmake header only library.

**Comment by hamishwillee** [comment]
Does this allow us easily to switch to development.xml, as discussed in https://github.com/PX4/PX4-Autopilot/pull/17158 ?

**Comment by JacobCrabill** [comment]
> Does this allow us easily to switch to development.xml, as discussed in https://github.com/PX4/PX4-Autopilot/pull/17158 ?

Yep, we just change the dialect that we build in the cmake file.

**Comment by dagar** [comment]
Already bumped to development.xml here. https://github.com/PX4/PX4-Autopilot/blob/d8a61480d60fdbb0366247c4707ae9db401ce489/src/modules/mavlink/CMakeLists.txt#L34

I started using ifdefs for the messages in development.xml so that standard still works. https://github.com/PX4/PX4-Autopilot/blob/d8a61480d60fdbb0366247c4707ae9db401ce489/src/modules/mavlink/mavlink_receiver.cpp#L199-L204

After https://github.com/PX4/PX4-Autopilot/pull/17100 we'll be able to optionally select the dialect per board variant. https://github.com/PX4/PX4-Autopilot/pull/17100

**Comment by hamishwillee** [comment]
FYI only, this will work as long as we include the headers in whatever the top level dialect is built (e.g. the ones in development if we build development). In theory you should be able to generate headers for development.xml but link against the headers in the generated common folder. That will not work due to a bug in the generators - the header for the top level dialect you built is the only one that has MAV_CMD definitions.

**Comment by dagar** [comment]
> FYI only, this will work as long as we include the headers in whatever the top level dialect is built (e.g. the ones in development if we build development). In theory you should be able to generate headers for development.xml but link against the headers in the generated common folder. That will not work due to a bug in the generators - the header for the top level dialect you built is the only one that has MAV_CMD definitions.
> 
> PS. And if we do not use this approach (i.e. this PR does not merge) then according to @julianoes there is some conflict with the prebuild libraries that still needs looking at:
> 
> ```
> ../../mavlink/include/mavlink/v2.0/development/development.h:345: error: "MAVLINK_MESSAGE_INFO" redefined [-Werror]
> ```

I believe that should all be fine. I've also updated the include path to point to the dialect, then in all the code it's now just `#include <mavlink.h>` instead of `#include <standard/mavlink.h>` (or whatever).

Does that cover everything?

File: .gitmodules (hunk 3 above)
Comment (bkueng) [reply .gitmodules]: `src/modules/mavlink/mavlink` is fine with me as well

**Comment by JacobCrabill** [comment]
Perfect!  I do a tiny amount of work, and then get the credit for a now much-larger PR and cleanup effort 😂 

Nice work @dagar - I think this is a really great piece of cleanup and streamlining work.

**Comment by JacobCrabill** [comment]
What about CI Docker containers?  Do they all contain the Python3 / Ubuntu dependencies to run mavgen and the xml checks?

**Comment by dagar** [comment]
> Perfect! I do a tiny amount of work, and then get the credit for a now much-larger PR and cleanup effort 😂
> 
> Nice work @dagar - I think this is a really great piece of cleanup and streamlining work.

Thanks Jacob, although I must admit I bit off slightly more than intended here now that I see the mess in submodule dependencies. The growing list of simulator submodules (gazebo, ignition, jsbsim, flightgear) were all looking for mavlink in firmware, so they'll all need to be updated once this settles (with or without the cmake header only library idea).

It seems pretty awkward to have these submodule dependencies either poke into the parent to find mavlink or look for it on the system. I think it will be easier to have them all generate mavlink directly.

**Comment by dagar** [comment]
> What about CI Docker containers? Do they all contain the Python3 / Ubuntu dependencies to run mavgen and the xml checks?

This part seems to be fine.

**Comment by julianoes** [comment]
> The growing list of simulator submodules (gazebo, ignition, jsbsim, flightgear) were all looking for mavlink in firmware, so they'll all need to be updated once this settles (with or without the cmake header only library idea).

They could be moved to firmware since they need to move in lockstep anyway, usually.

**Comment by RomanBapst** [comment]
@bkueng @dagar On my computer I get an average of 3.5 seconds longer build time with this PR (target fmuv5, 4 builds with make clean / make distclean in between).
Is there more testing required regarding build time?

**Comment by dagar** [comment]
Update for sitl_gazebo. https://github.com/PX4/PX4-SITL_gazebo/pull/822

**Comment by dagar** [comment]
Next step is make the mavlink dialect configurable with kconfig.

**Pull request merged**